pbm/ctrl/recv_test.go

```diff
@@ -0,0 +1,142 @@
+package ctrl
+
+import (
+	"context"
+	"log"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/modules/mongodb"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+
+	"github.com/percona/percona-backup-mongodb/pbm/connect"
+)
+
+var (
+	mClient    *mongo.Client
+	connClient connect.Client
+)
+
+func TestMain(m *testing.M) {
+	ctx := context.Background()
+	mongodbContainer, err := mongodb.Run(ctx, "perconalab/percona-server-mongodb:8.0.4-multi")
+	if err != nil {
+		log.Fatalf("error while creating mongo test container: %v", err)
+	}
+	connStr, err := mongodbContainer.ConnectionString(ctx)
+	if err != nil {
+		log.Fatalf("conn string error: %v", err)
+	}
+	mClient, err = mongo.Connect(ctx, options.Client().ApplyURI(connStr))
+	if err != nil {
+		log.Fatalf("mongo client connect error: %v", err)
+	}
+
+	connClient = connect.UnsafeClient(mClient)
+
+	code := m.Run()
+
+	err = mClient.Disconnect(ctx)
+	if err != nil {
+		log.Fatalf("mongo client disconnect error: %v", err)
+	}
+	if err := testcontainers.TerminateContainer(mongodbContainer); err != nil {
+		log.Fatalf("failed to terminate container: %s", err)
+	}
+
+	os.Exit(code)
+}
+
+func TestListenCmdFiltersDuplicate(t *testing.T) {
+	ctx := context.Background()
+	coll := connClient.CmdStreamCollection()
+	_ = coll.Drop(ctx)
+
+	restoreTS := time.Now().UTC().Unix()
+	backupTS := restoreTS + 1
+
+	restoreDoc := bson.D{
+		{"_id", primitive.NewObjectID()},
+		{"cmd", "restore"},
+		{"ts", restoreTS},
+	}
+	backupDoc := bson.D{
+		{"_id", primitive.NewObjectID()},
+		{"cmd", "backup"},
+		{"ts", backupTS},
+	}
+
+	if _, err := coll.InsertOne(ctx, restoreDoc); err != nil {
+		t.Fatalf("insert restore: %v", err)
+	}
+	if _, err := coll.InsertOne(ctx, backupDoc); err != nil {
+		t.Fatalf("insert backup: %v", err)
+	}
+
+	cmdC, errC := ListenCmd(ctx, connClient, make(chan struct{}))
+
+	var got []Cmd
+	for len(got) < 2 { // want only restore + backup
+		select {
+		case cmd := <-cmdC:
+			got = append(got, cmd)
+		case err := <-errC:
+			t.Fatalf("listener error: %v", err)
+		}
+	}
+
+	if len(got) != 2 || got[0].Cmd != "restore" || got[1].Cmd != "backup" {
+		t.Errorf("duplicate leaked through: %v", got)
+	}
```

Comment:
I would expand the same test with following cases:
- 3 commands within the same second
- 3 commands in T1, 2 commands in T2, 3 commands in T3